internal/configs/resource.go

```diff
@@ -816,12 +816,22 @@ func (r *ProviderConfigRef) InstanceValidation(blockType string, isInstanced boo
 			})
 		}
 		if !isInstanced {
-			diags = append(diags, &hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  summary,
-				Detail:   "An instance key can be specified only for a provider configuration that uses for_each.",
-				Subject:  r.KeyExpression.Range().Ptr(),
-			})
+			if !hasConfig {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  summary,
+					Detail:   fmt.Sprintf("An instance key can only be used with a provider configured with for_each. No provider configuration was detected for %q", r.String()),
+					Subject:  r.KeyExpression.Range().Ptr(),
```

Comment:
Oh, also:

This doesn't seem like the right range for this error message. I think instead we want `hcl.RangeBetween(r.NameRange, *r.AliasRange)` so that it's the `postgres.by_db` part that gets highlighted in the diagnostic, rather than the `[each.key]` part.